### ISSUE TITLE
Consistent per-meeting CREATE parameter for enabled/disabled features

### ISSUE THREAD

**Issue by antobinary** [bigbluebutton/bigbluebutton#14203]
**Is your feature request related to a problem? Please describe.**
Implement a consistent way to control the list of enabled vs disabled features on per-meeting level.

**Describe the solution you'd like**

Add a new parameter with a comma separated values like this `disabledFeatures=video,audio`.

Caveats:
 - the widest range of possibly desired features would need to be all enabled in configuration (settings.yml / bigbluebutton.props ) 
 - if passing on create to disable feature X, and on join to enable it for the user, the feature will be disabled
 - if we do not mention X in the list of disabled features on create, userdata disable should still disable it for the user on client side

(At time of creation of this issue, the following members have shared their initial thoughts: @kepstin  @GhaziTriki @ritzalam @gustavotrott @pedrobmarin )

**Describe alternatives you've considered**
The way we do it now is a mix of
- no way to disable features like Chat, Polling, (see pending pull requests https://github.com/bigbluebutton/bigbluebutton/pull/14152 https://github.com/bigbluebutton/bigbluebutton/pull/11384 )
- create parameter: `learningDashboardEnabled`
- join parameters which are passed for **all** users: `userdata-bbb_enable_screen_sharing=`
and it's inconsistent, we receive enhancement requests for per-meeting control of additional features, ...

**Additional context**

**Full list of features to be handled:**

audio (disabled means no action buttons about it, no effect of parameters like auto join, skip check, etc)
video (or webcams?)
screenshare (disabled means no action buttons about it)
chat
polling (disabled means omitted from the client panels, from the '+' action bar menu)
selectRandomUser (disabled means omitted from the client panels, from the '+' action bar menu)
breakoutRooms (disabled means omitted from the client panels, gear menu, ...)
captions (disabled means omitted from the client panels, from the '+' action bar menu)
sharedNotes (disabled means omitted from the client panels)
layouts (disabled means stuck with the default layout, no UX to modify)
(please comment for ones missed)

**Shortcuts**

Shortcuts about disabed features won't be displayed and won't function.
We may need to inform that the shortcut is not working because the feature is off (for accessibility purposes)

**Rare cases to test**

* allow chat only among moderators (i.e. enabled per-meeting on create but disabled on join for viewers)
* ...?

**Documentation**

Document the new API parameter
Document the beginning of deprecation for `learningDashboardEnabled` and other per-meeting alternatives. I think we need to keep the existing userdata_X parameters (see caveats above).

**Please include in your comments:**

- are you aware of any use case where you'd only disable a feature for specific users (note, this resembles lock settings)

**Comment by gustavotrott** [comment]
I like the idea of let API only disable features, instead of disable **and enable**.

So, if any instance set the feature as disabled, it will always have superiority over enabled.
And features will be available only if they are enabled in config, and absent in both requests CREATE and JOIN.

We would not have conflicts between CREATE and JOIN, once the user will receive the sum of both `disabledFeatures`.

In short, I like this idea:
![image](https://user-images.githubusercontent.com/5660191/151054986-8aa76779-fdde-4b98-a97e-597f67185712.png)

**Comment by pedrobmarin** [comment]
I like the idea of disabling a feature in a particular meeting but I'm not a big fan of changing it for a single user. I think it can bring a big load of complexity codewise and can be misleading from the user perspective.

**Comment by antobinary** [comment]
I am thinking whether we still need to keep `userdata-bbb_enable_screen_sharing` once we support `disabledFeatures=screenshare`

In my experience we either disable the feature for the whole meeting, or we don't disable at all.
What do you think?

**Comment by TiagoJacobs** [comment]
In my opinion, we should support both on create and on join.

It would be a decision of the 3rd party application to use one or other approach.

The reason to keep it on join is because the 3rd party application can apply new configuration without recreating the session (the user will only need to rejoin).

PS: to keep rules simple to understand, we could prohibit the usage of both, so once you use `disabledFeatures` on create, the join will ignore+warn if you try to use it again.

**Comment by gustavotrott** [comment]
At the moment, the feature disabling is already being applied on user level, right?

I think we could keep it always at user level. And the only difference is that users could also inherit the `disabledFeatures` of the meeting.

> PS: if it's going to make simpler, we could prohibit the usage of both together: so once you use disabledFeatures on create, the join will ignore+warn if you try to use it again.

I can't see a good reason to avoid the usage in both (CREATE + JOIN). Once it doesn't seem to be complex to implement and could possibly be useful.

**Comment by pedrobmarin** [comment]
So, let me get things right.

I'm imagining a meeting were some user is able to share his video and another isn't because he got a `disabledFeatures=video` at his join URL. If I'm this user the first thing I do is open a ticket saying that something was wrong at the meeting because people were able to share their camera and I wasn't. How this will differentiate from the lock settings? How many more use cases we will need to test before a new release? Is this going to fully replace `userdata`?

**Comment by gustavotrott** [comment]
Hey, good concerns @pedrobmarin.

My thoughts about it are:

> Users opening tickets: 

I can't see a scenario where a given user would have some feature disabled randomly. I mean, if the video is disabled it would probably be applied for all similar users. I guess normally for hierarchy reasons. Or at least the moderator would know why it is disabled only for some users.

> How many more use cases we will need to test before a new release?

You are right, it would probably increase the test scenarios and I'm not able to measure it and comment if it worth to reduce options in favor of testings.

> How this will differentiate from the lock settings?
Is this going to fully replace userdata?

I didn't think deeper about the details of the implementation.
But I can see a big advantage in being able to set `disabledFeatures` as a CREATE param. Currently using `userdata` is is possible to set only in the JOIN request.

I'm sensitive about the @TiagoJacobs's point: If the `disabledFeatures` be accepted only on CREATE request. The only way to change something is recreating the meeting. Meanwhile if they prefers to disable features on JOIN request, they would be able to change options easily. This decision would be up to the 3rd party.

Anyway, I'm just trying to throw up some points to think about.
But I know you guys know much more use cases! I also like the idea of "keep things simple".

**Comment by pedrobmarin** [comment]
Maybe we could experiment building this over the create call and give some time to collect feedback from the community before taking the next step.

**Comment by TiagoJacobs** [comment]
> How this will differentiate from the lock settings
Lock settings are restricted to viewers and provides a way for end-users to enable/disable permissions. The audience of this parameter on join would be different: the 3rd party application.

> Maybe we could experiment building this over the create call and give some time to collect feedback from the community before taking the next step
It looks wise, but will lead to rework, that's why it's good to discuss if we need it or not, so the data structures can be created accordingly.

@antobinary , in your initial requirement, there is a section "rare cases to test", if we decide to go without the "per user" configuration, you would be removing this requirement?

**Comment by antobinary** [comment]
> @antobinary , in your initial requirement, there is a section "rare cases to test", if we decide to go without the "per user" configuration, you would be removing this requirement?

I would definitely be dropping some cases and we'd be closer to the KISS principle. There are just too many places where a configuration may occur:
bbb properties file, settings.yml file, create param, [join param], lock settings, possible logic depending on role, 
Even dropping the userdata parameter for disabling a feature, we'd still benefit from listing edge cases to test...

**Comment by gustavotrott** [comment]
I sent an initial proposal in #14293

**Comment by antobinary** [comment]
Let's iterate on the full list:

audio -> ____ (suggestion welcome)
webcams
screenshare
chat
polls
selectRandomUser
breakoutRooms
captions
sharedNotes
layouts 
(please comment for ones missed)

**Comment by pedrobmarin** [comment]
learningDashboard?

**Comment by liske** [comment]
paging toggle

**Comment by gustavotrott** [comment]
virtualBackgrounds?
checkAudio?
record?
importBreakoutAnnotations? (future)

Is something related with presentation or whiteboard/annotations intended? or officeToPdfConversion?

**Comment by maths22** [comment]
Can public chat and 1-1 chat be separated out?  I have a use case where 1-1 chat is bad but public chat is fine, and I'd imagine some other people might find that useful.

**Comment by g-v-egidy** [comment]
Since there is discussion about if it should be possible to enable/disable features per user via the API - I would prefer if that would be possible.

I currently have several cases where I forbid everyone but one to share video, also the presentation is auto-hidden. The users without right to share video also have set `userdata-bbb_auto_join_audio: false` and `userdata-bbb_skip_check_audio: true`. This effectively turns BBB into a video sharing solution where the regular participants join and don't have to interact with any BBB controls or popups and just get video+audio presented directly after they log in. Also they don't have any unnecessary UI elements or means to misconfigure anything. I would prefer to keep this supported.

I think the "disable only" way of controlling it, as suggested by @gustavotrott in https://github.com/bigbluebutton/bigbluebutton/issues/14203#issuecomment-1021584847 would be a consistent way to do it.

**Comment by antobinary** [comment]
> Can public chat and 1-1 chat be separated out?

@gustavotrott if we allow for `publicChat` vs `privateChat` and handle separately, do we need to jump through hoops to support a use case like `publicChat` disabled and `privateChat` enabled?

**Comment by gustavotrott** [comment]
> > Can public chat and 1-1 chat be separated out?
> 
> @gustavotrott if we allow for `publicChat` vs `privateChat` and handle separately, do we need to jump through hoops to support a use case like `publicChat` disabled and `privateChat` enabled?

I think it is very attainable, once we already have `lockSettings.disablePrivateChat` and `lockSettings.disablePublicChat`.
It will require some work, but doesn't seem to be painful :sweat_smile:

**Comment by antobinary** [comment]
Quite a few of these were handled as part of BBB 2.5.
Maybe @gustavotrott could link to a summary?

Let's do the rest in 2.6 (wraping up 2.5 to be able to transition to beta stage)

**Comment by gustavotrott** [comment]
> Can public chat and 1-1 chat be separated out? I have a use case where 1-1 chat is bad but public chat is fine, and I'd imagine some other people might find that useful.

We will address it to 2.6, but in the meanwhile you can achieve it using `lockSettingsDisablePrivateChat=true`
Take a look in https://docs.bigbluebutton.org/dev/api.html#api-calls

At least viewers will have private chat disabled.

**Comment by gustavotrott** [comment]
**Done 2.5:**
breakoutRooms ✅
captions ✅
chat ✅
externalVideos ✅
layouts ✅
learningDashboard ✅
polls ✅
screenshare ✅
sharedNotes ✅
virtualBackgrounds ✅

**Pending 2.6:**
checkAudio
importBreakoutAnnotations
microphone
officeToPdfConversion
pagingToggle
publicChat/privateChat
recording
selectRandomUser
webcam

**Comment by antobinary** [comment]
Related: #15346 #15347 #15348

**Comment by lonesomewalker** [comment]
I am not so sure about this:

Done 2.6:
downloadPresentationWithAnnotations

We currently experience the problem that disabling this feature does nothing in the frontend, either by globally setting this in bbb-web.properties or by additional parameter in create url. :-(

**Comment by gustavotrott** [comment]
> I am not so sure about this:
> 
> Done 2.6: downloadPresentationWithAnnotations
> 
> We currently experience the problem that disabling this feature does nothing in the frontend, either by globally setting this in bbb-web.properties or by additional parameter in create url. :-(

Hey @lonesomewalker, thanks for reporting. It was supposed to be working.
Could you please open an Issue reporting this problem?
This feature is the button:
<table><tr><td>

![image](https://github.com/bigbluebutton/bigbluebutton/assets/5660191/437d8ef6-a204-409e-9624-2f64c8610ab4)

</td></tr></table>

**Comment by antobinary** [comment]
(We've been adding to the list and most new functionality gets a disabledFeatures property upon being added)
